### PR TITLE
compiler/next: MultiDecl

### PR DESCRIPTION
**Pull request by mppf**
Add MultiDecl AST, parse it, and test for it.

MultiDecl represents a declaration for
multiple variables.

E.g.

``` chpl

    var a, b:int, c, d = 1;
    var x: int, y = 3, z: real;
```

Each of the lines above is represented by a MultiDecl containing a
list of VariableDecls.  Note that the initial value and/or type is inferred
from later declarations.

Since the MultiDecl does not declare a Sym itself, it is not a
subclass of Decl. Rather, it contains several VariableDecls.

Besides the above, this PR comments out the SimpleType AST node from 
ASTClassesList because I am currently thinking we won't need it (because
we don't ever define things like `int` in the module code). However I am
leaving it there as a comment in case it turns out to be necessary.

Reviewed by @vasslitvinov - thanks!

### CODE REVIEW

File: compiler/next/include/chpl/uast/ASTClassesList.h [hunk 1]
```diff
@@ -130,7 +131,7 @@ AST_BEGIN_SUBCLASSES(Sym)              // old AST: Symbol
 
   AST_BEGIN_SUBCLASSES(TypeSym)        // old AST: TypeSymbol/Type
     AST_NODE(Enum)                     // old AST: EnumType
-    AST_NODE(SimpleType)               //
+    //AST_NODE(SimpleType)             //
```
Comment (vasslitvinov) [inline compiler/next/include/chpl/uast/ASTClassesList.h:134]: If commenting out `SimpleType` is intended, please explain in the PR message, esp. why still keep this line in the source code.

File: compiler/next/include/chpl/uast/MultiVarDeclaration.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  ASTListIteratorPair<Expression> stmts() const {
+    return ASTListIteratorPair<Expression>(children_.begin(), children_.end());
+  }
+
+  /**
+   Return the number of VariableDecls and Comments contained.
+   */
+  int numStmts() const {
+    return this->numChildren();
+  }
```
Comment (vasslitvinov) [inline compiler/next/include/chpl/uast/MultiVarDeclaration.h:82]: Calling these things "statements" is not intuitive at a first glance.

I'd call these methods `numDeclsComments()` and `declOrComment(i)`.

**Review by vasslitvinov — APPROVED** [review]
Need to update Decl.h to indicate that it is for declaring a single symbol and point to MultiVarDeclaration for the multiple-symbol case.

File: compiler/next/include/chpl/uast/MultiVarDeclaration.h (hunk 2 above)
Comment (mppf) [reply compiler/next/include/chpl/uast/MultiVarDeclaration.h:82]: I have `numDeclOrComments` and `declOrComment` for now.

Comment (mppf) [reply compiler/next/include/chpl/uast/MultiVarDeclaration.h:82]: I'm going to look at adding an iterator to get the declared Syms and ignore comments. I'll add that in #17691.

**Comment by vasslitvinov** [comment]
> Need to update Decl.h to indicate that it is for declaring a single symbol and point to MultiVarDeclaration for the multiple-symbol case.

As mentioned in https://github.com/chapel-lang/chapel/pull/17658#discussion_r631843510, we want to support tuple declarations like `var (a,b) = ....;` in which case `Decl` does not apply either as there is no single symbol being declared.

I don't want to end up with a `Decl` and several other decl types that do not inherit from `Decl`. Instead, perhaps rename `Decl` into `SingleSymbolDecl` and have `Decl` have children like `MultiVarDeclaration` and `TupleDeclaration`.

Then again, how about the case `var (a,b) = ..., (c,d) = ...;` ?

**Comment by dlongnecke-cray** [comment]
Why not `SingleVarDecl` and `MultiVarDecl`? Just wondering why `SingleSymbol` is the odd one out.

That's just a nit about names, but I think this suggestion is good. 👍

**Comment by vasslitvinov** [comment]
> Just wondering why `SingleSymbol` is the odd one out.

Because currently (i.e., last I looked) `Decl` contained a method that returned THE symbol it is declaring. This led me to the name "SingleSymbol". However, I do not expect a "SingleSymbol" parent to be all that useful.

"SingleVarDecl" is surely needed, however I suggest we do not need "Single" in its name.

**Comment by mppf** [comment]
> However, I do not expect a "SingleSymbol" parent to be all that useful.

I am thinking that we need a parent that does this because:
 1. I'm imagining FunctionDecl and related things to inherit from it and
 2. I'm imagining that the MultiVarDecl and TupleDecl will contain SingleSymbolDecls (whatever we call it)

**Comment by vasslitvinov** [comment]
To clarify, there are two "single" decls on the table:

* `SingleVarDecl` - to represent things like `var x = ...;` Certainly needed. I suggest leaving "single" out of its name.

* `SingleSymbolDecl` - a parent of FunctionDecl and similar. While aesthetically pleasing, I am not seeing what code it would factor out. Yes, it would provide a method for THE symbol being declared. However, this will not be needed if `Decl` provides an iterable over all its symbols.

**Comment by mppf** [comment]
We had an off-issue discussion. Per our discussion I'm planning to change the Decl inheritance hierarchy to this:
```
Decl
  MultiDecl
  TupleDecl  // these contain several SymDecls
  
  SymDecl // declares a single Sym
    FormalDecl
    VariableDecl
    FunctionDecl
```

**Pull request merged**